pkg/server/server.go

```diff
@@ -274,6 +336,51 @@ func (h *handler) authenticateEndpoint(w http.ResponseWriter, req *http.Request)
 	})
 }
 
+func (h *handler) getPrivateDNSName(id string) (string, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	name, ok := h.nodeNameCache[id]
+	if ok {
+		return name, nil
+	}
+	return "", errors.New("instance id not found")
+}
+
+func (h *handler) setPrivateDNSName(id string, privateDNSName string) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	h.nodeNameCache[id] = privateDNSName
+}
+
+func (h *handler) getPrivateDNSFromEC2(id string) (string, error) {
+	privateDNSName, err := h.getPrivateDNSName(id)
+	if err == nil {
+		return config.NodeNamePrefix + privateDNSName, nil
+	}
+
+	// Look up instance from EC2 API
+	instanceIds := []*string{&id}
+	ec2Service := ec2.New(h.sess)
+	output, err := ec2Service.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: instanceIds,
```

Comment:
Consider using privateDnsName filter 

https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#DescribeInstancesInput

Reply:
All I know is the instanceId when making the request, so I wouldn't be able to use the filter, unless I don't understand how it works?

Reply:
I made a mistake while reading through the API documentation and presumed it's an output filter. Ignore this comment.